aten/src/ATen/native/cuda/SegmentReduce.cu

```diff
@@ -70,21 +71,192 @@ __global__ static void post_sum_div_kernel(
   }
 }
 
+template <typename scalar_t>
+__global__ static void segment_reduce_forward_kernel(
+    SegmentReductionType reduction,
+    scalar_t* output_data,
+    scalar_t* values_data,
+    const int64_t* lengths_data,
+    const int64_t* lengths_cumsum_data,
+    const int64_t segment_count,
+    const int64_t stride_count,
+    bool is_initial_set,
+    scalar_t initial_value) {
+  int64_t idx = blockIdx.x * blockDim.x + threadIdx.x;
+  int64_t row_id = idx / stride_count;
+  int64_t lane_id = idx % stride_count;
+  if (idx >= (segment_count * stride_count)) {
+    return;
+  }
+  int64_t offset_start = lengths_cumsum_data[row_id];
+  int64_t offset_end = lengths_cumsum_data[row_id + 1];
+
+  // ===== step2: apply reduction
+  for (int64_t j = offset_start; j < offset_end; ++j) {
+    int64_t starting_index = (j * stride_count) + lane_id;
+    const auto data = values_data[starting_index];
+    // TODO: There is no need to branch with every element
+    if (reduction == SegmentReductionType::MAX) {
```

Comment:
if perf is important, you can make `reduction` a template parameter, and branch will be eliminated at compile time